src/Language/DifferentialDatalog/Validate.hs

```diff
@@ -356,8 +357,16 @@ atomValidate d ctx atom = do
 -- Validate an RHS term of a rule.  Once all RHS and LHS terms have been
 -- validated, it is safe to call 'ruleValidateExpressions'.
 ruleRHSValidate :: (MonadError String me) => DatalogProgram -> Rule -> RuleRHS -> Int -> me ()
-ruleRHSValidate d rl (RHSLiteral _ atom) idx =
+ruleRHSValidate d rl (RHSLiteral polarity atom) idx = do
     atomValidate d (CtxRuleRAtom rl idx) atom
+    let rel = getRelation d $ atomRelation atom
+    when (rulePrefixIsStream d rl idx) $ check d (relSemantics rel /= RelStream) (pos atom)
+        $ "Illegal join of two streams: stream relation '" ++ name rel ++ "' and the stream produced by the prefix of the rule before this literal. Streams can only be joined with regular non-stream relations."
+    when (relIsStream rel) $ check d polarity (pos atom)
+        $ "Illegal negation of a stream relation '" ++ name rel ++ "'. Only non-stream relations can be negated."
+    -- This is an implementation constraint.  We should be able to antijoin streams.
+    when (polarity == False) $ check d (not $ rulePrefixIsStream d rl idx) (pos atom)
+        $ "Unsupported antijoin. The prefix of the rule before this literal produces a stream. Antijoins with streams are currently not supported."
```

Comment:
Nit, the wording is a bit harsh, maybe something like "attempted to join streams {stream1} and {stream2} which is not currently supported" would be better